src/utils/editorConfig.ts

```diff
@@ -0,0 +1,143 @@
+import { parseString, KnownProps } from "editorconfig";
+import { Minimatch } from "minimatch";
+
+export interface EditorConfig {
+    root: boolean;
+    sections: EditorConfigSection[];
+}
+
+export interface EditorConfigSection {
+    pattern: string;
+    props: EditorConfigSectionProps;
+}
+
+export type EditorConfigSectionProps = KnownProps;
+
+// Find .editorconfig files in parent directories
+export function getEditorConfigPaths(paths: string[], path?: string) {
+    if (!path) {
+        return paths.filter((path) => path.split("/").pop() === ".editorconfig");
+    }
+
+    const editorConfigPaths: string[] = [];
+    const pathComponents = path.split("/");
+
+    while (pathComponents.length) {
+        const parentPath = pathComponents.join("/");
+        const parentEditorConfigPath = `${parentPath}/.editorconfig`;
+
+        if (paths.indexOf(`${parentPath}/.editorconfig`) > -1) {
+            editorConfigPaths.push(parentEditorConfigPath);
+        }
+
+        pathComponents.pop();
+    }
+
+    if (paths.indexOf(".editorconfig") > -1) {
+        editorConfigPaths.push(".editorconfig");
+    }
+
+    return editorConfigPaths;
+}
+
+// Parse .editorconfig text
+// https://editorconfig-specification.readthedocs.io/en/latest/
+export function parseEditorConfig(data: string) {
+    const parsedConfig: EditorConfig = {
+        root: false
+      , sections: []
+    };
+
+    for (const [ pattern, props ] of parseString(data)) {
+        // Preamble has null pattern
+        if (pattern === null) {
+            if (props.root === "true") {
+                parsedConfig.root = true;
+            } else if (props.root === "false") {
+                parsedConfig.root = false;
+            }
+
+            continue;
+        }
+
+        const parsedProps: KnownProps = {};
+
+        if ("indent_style" in props) {
+            if (props.indent_style === "tab" || props.indent_style === "space") {
+                parsedProps.indent_style = props.indent_style;
+            }
+        }
+        if ("indent_size" in props) {
+            if (props.indent_size === "tab") {
+                parsedProps.indent_size = "tab";
+            } else {
+                const indentSize = Number(props.indent_size);
+                if (!isNaN(indentSize)) {
+                    parsedProps.indent_size = indentSize;
+
+                    // tab_width defaults to indent_size
+                    if (!("tab_width" in props)) {
+                        parsedProps.tab_width = indentSize;
```

Comment:
On a more serious note, this code can be simplified (so that the comment is really not necessary):
```suggestion
                        parsedProps.tab_width = parsedProps.tab_width || indentSize;
```

Or actually (if the target supports it) that brand new [?? operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator):

```suggestion
                        parsedProps.tab_width = parsedProps.tab_width ?? indentSize;
```

Reply:
careful there:
- props and parsedProps are two different things.
- in your first suggestion, if tab_width is 0 or emptystring, they would be ignored.
- in your second suggestion, only null or undefined would fallback to indentsize.
- looking at the following lines of code, it converts props.tab_width from string to number.